File: src/exchange/paper.py
"""
WhaleTrader - Paper Trading Engine
Simulates trades without real money. Default mode for safety.
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional
from enum import Enum


class OrderSide(Enum):
    BUY = "buy"
    SELL = "sell"


class OrderStatus(Enum):
    PENDING = "pending"
    FILLED = "filled"
    CANCELLED = "cancelled"
    REJECTED = "rejected"


@dataclass
class Order:
    id: str
    asset: str
    side: OrderSide
    quantity: float
    price: float
    status: OrderStatus = OrderStatus.PENDING
    filled_price: Optional[float] = None
    filled_at: Optional[datetime] = None
    created_at: datetime = field(default_factory=datetime.now)
    reason: str = ""  # Why the agent made this trade


@dataclass
class Position:
    asset: str
    quantity: float
    avg_entry_price: float
    current_price: float = 0.0
    opened_at: datetime = field(default_factory=datetime.now)

    @property
    def market_value(self) -> float:
        return self.quantity * self.current_price

    @property
    def unrealized_pnl(self) -> float:
        return (self.current_price - self.avg_entry_price) * self.quantity

    @property
    def unrealized_pnl_pct(self) -> float:
        if self.avg_entry_price == 0:
            return 0
        return (self.current_price - self.avg_entry_price) / self.avg_entry_price


class PaperTradingEngine:
    """
    Paper trading engine that simulates real trading without actual money.
    Tracks positions, P&L, and order history.
    """

    def __init__(self, initial_capital: float = 10000.0):
        self.initial_capital = initial_capital
        self.cash = initial_capital
        self.positions: dict[str, Position] = {}
        self.order_history: list[Order] = []
        self.trade_log: list[dict] = []
        self._order_counter = 0

    @property
    def total_value(self) -> float:
        """Total portfolio value (cash + positions)"""
        positions_value = sum(p.market_value for p in self.positions.values())
        return self.cash + positions_value

    @property
    def total_pnl(self) -> float:
        """Total profit/loss since inception"""
        return self.total_value - self.initial_capital

    @property
    def total_pnl_pct(self) -> float:
        """Total P&L as percentage"""
        return self.total_pnl / self.initial_capital

    def update_prices(self, prices: dict[str, float]):
        """Update current prices for all positions"""
        for asset, price in prices.items():
            if asset in self.positions:
                self.positions[asset].current_price = price

    def place_order(self, asset: str, side: OrderSide, quantity: float,
                    current_price: float, reason: str = "") -> Order:
        """
        Place a paper trade order. Executes immediately at current price.
        
        Args:
            asset: Asset symbol
            side: BUY or SELL
            quantity: Amount to trade
            current_price: Current market price
            reason: AI agent's reasoning for the trade
            
        Returns:
            Filled Order object
        """
        self._order_counter += 1
        order_id = f"PAPER-{self._order_counter:06d}"

        order = Order(
            id=order_id,
            asset=asset,
            side=side,
            quantity=quantity,
            price=current_price,
            reason=reason,
        )

        # Validate order
        if side == OrderSide.BUY:
            cost = quantity * current_price
            if cost > self.cash:
                order.status = OrderStatus.REJECTED
                order.reason = f"Insufficient cash. Need ${cost:,.2f}, have ${self.cash:,.2f}"
                self.order_history.append(order)
                return order
        elif side == OrderSide.SELL:
            if asset not in self.positions or self.positions[asset].quantity < quantity:
                available = self.positions[asset].quantity if asset in self.positions else 0
                order.status = OrderStatus.REJECTED
                order.reason = f"Insufficient position. Want to sell {quantity}, have {available}"
                self.order_history.append(order)
                return order

        # Execute order (paper trades fill immediately)
        order.status = OrderStatus.FILLED
        order.filled_price = current_price
        order.filled_at = datetime.now()

        if side == OrderSide.BUY:
            self._execute_buy(asset, quantity, current_price)
        else:
            self._execute_sell(asset, quantity, current_price)

        self.order_history.append(order)
        self._log_trade(order)

        return order

    def _execute_buy(self, asset: str, quantity: float, price: float):
        """Execute a buy order"""
        cost = quantity * price
        self.cash -= cost

        if asset in self.positions:
            pos = self.positions[asset]
            total_cost = (pos.avg_entry_price * pos.quantity) + cost
            pos.quantity += quantity
            pos.avg_entry_price = total_cost / pos.quantity
            pos.current_price = price
        else:
            self.positions[asset] = Position(
                asset=asset,
                quantity=quantity,
                avg_entry_price=price,
                current_price=price,
            )

    def _execute_sell(self, asset: str, quantity: float, price: float):
        """Execute a sell order"""
        proceeds = quantity * price
        self.cash += proceeds

        pos = self.positions[asset]
        realized_pnl = (price - pos.avg_entry_price) * quantity

        pos.quantity -= quantity
        pos.current_price = price

        if pos.quantity <= 0.0001:  # Close position (float precision)
            del self.positions[asset]

        return realized_pnl

    def _log_trade(self, order: Order):
        """Log trade for analytics"""
        self.trade_log.append({
            "order_id": order.id,
            "asset": order.asset,
            "side": order.side.value,
            "quantity": order.quantity,
            "price": order.filled_price,
            "timestamp": order.filled_at.isoformat(),
            "reason": order.reason,
            "portfolio_value": self.total_value,
            "cash": self.cash,
        })

    def get_portfolio_summary(self) -> dict:
        """Get current portfolio summary"""
        positions_data = {}
        for asset, pos in self.positions.items():
            positions_data[asset] = {
                "quantity": pos.quantity,
                "avg_entry": pos.avg_entry_price,
                "current_price": pos.current_price,
                "market_value": pos.market_value,
                "unrealized_pnl": pos.unrealized_pnl,
                "unrealized_pnl_pct": f"{pos.unrealized_pnl_pct:.2%}",
            }

        return {
            "initial_capital": self.initial_capital,
            "cash": self.cash,
            "positions_value": sum(p.market_value for p in self.positions.values()),
            "total_value": self.total_value,
            "total_pnl": self.total_pnl,
            "total_pnl_pct": f"{self.total_pnl_pct:.2%}",
            "num_positions": len(self.positions),
            "num_trades": len(self.trade_log),
            "positions": positions_data,
        }

    def __repr__(self):
        return (f"<PaperTradingEngine: ${self.total_value:,.2f} "
                f"({self.total_pnl_pct:+.2%}) | "
                f"{len(self.positions)} positions | "
                f"{len(self.trade_log)} trades>")

File: engine/src/exchange/paper.rs
//! WhaleTrader Paper Trading Exchange
//! Simulates real exchange behavior without risking money.

use crate::core::types::*;
use chrono::Utc;
use rust_decimal::Decimal;
use std::collections::HashMap;
use tracing::{info, warn};
use uuid::Uuid;

/// Paper trading engine — simulates an exchange
pub struct PaperExchange {
    cash: Decimal,
    initial_capital: Decimal,
    positions: HashMap<String, Position>,
    orders: Vec<Order>,
    trade_count: u64,
}

impl PaperExchange {
    pub fn new(initial_capital: Decimal) -> Self {
        info!("Paper exchange initialized with ${} capital", initial_capital);
        Self {
            cash: initial_capital,
            initial_capital,
            positions: HashMap::new(),
            orders: Vec::new(),
            trade_count: 0,
        }
    }

    pub fn cash(&self) -> Decimal {
        self.cash
    }

    pub fn total_value(&self) -> Decimal {
        let pos_value: Decimal = self.positions.values().map(|p| p.market_value()).sum();
        self.cash + pos_value
    }

    pub fn total_pnl(&self) -> Decimal {
        self.total_value() - self.initial_capital
    }

    pub fn positions(&self) -> &HashMap<String, Position> {
        &self.positions
    }

    /// Update market prices for all positions
    pub fn update_price(&mut self, instrument: &str, price: Decimal) {
        if let Some(pos) = self.positions.get_mut(instrument) {
            pos.current_price = price;
            pos.unrealized_pnl = (price - pos.avg_entry_price) * pos.quantity;
            pos.updated_at = Utc::now();
        }
    }

    /// Execute a market order immediately at given price
    pub fn execute_market_order(
        &mut self,
        instrument: &str,
        side: Side,
        quantity: Decimal,
        price: Decimal,
        reason: &str,
    ) -> Result<Order, String> {
        let mut order = Order::market(instrument, side, quantity, reason);

        match side {
            Side::Buy => {
                let cost = quantity * price;
                if cost > self.cash {
                    order.status = OrderStatus::Rejected;
                    warn!(
                        "Order rejected: insufficient cash. Need {}, have {}",
                        cost, self.cash
                    );
                    return Err(format!(
                        "Insufficient cash: need ${}, have ${}",
                        cost, self.cash
                    ));
                }

                self.cash -= cost;
                self.trade_count += 1;

                if let Some(pos) = self.positions.get_mut(instrument) {
                    // Average up/down
                    let total_cost =
                        (pos.avg_entry_price * pos.quantity) + cost;
                    pos.quantity += quantity;
                    pos.avg_entry_price = total_cost / pos.quantity;
                    pos.current_price = price;
                    pos.updated_at = Utc::now();
                } else {
                    self.positions.insert(
                        instrument.to_string(),
                        Position {
                            instrument: instrument.to_string(),
                            side: Side::Buy,
                            quantity,
                            avg_entry_price: price,
                            current_price: price,
                            unrealized_pnl: Decimal::ZERO,
                            realized_pnl: Decimal::ZERO,
                            opened_at: Utc::now(),
                            updated_at: Utc::now(),
                        },
                    );
                }

                order.status = OrderStatus::Filled;
                order.filled_quantity = quantity;
                order.avg_fill_price = Some(price);
                order.updated_at = Utc::now();
            }
            Side::Sell => {
                let pos = self.positions.get(instrument);
                if pos.is_none() || pos.unwrap().quantity < quantity {
                    order.status = OrderStatus::Rejected;
                    let available = pos.map_or(Decimal::ZERO, |p| p.quantity);
                    return Err(format!(
                        "Insufficient position: want to sell {}, have {}",
                        quantity, available
                    ));
                }

                let proceeds = quantity * price;
                self.cash += proceeds;
                self.trade_count += 1;

                let pos = self.positions.get_mut(instrument).unwrap();
                let realized = (price - pos.avg_entry_price) * quantity;
                pos.realized_pnl += realized;
                pos.quantity -= quantity;
                pos.current_price = price;
                pos.updated_at = Utc::now();

                // Remove position if fully closed
                if pos.quantity <= Decimal::new(1, 8) {
                    // ~0 threshold
                    self.positions.remove(instrument);
                }

                order.status = OrderStatus::Filled;
                order.filled_quantity = quantity;
                order.avg_fill_price = Some(price);
                order.updated_at = Utc::now();
            }
        }

        info!(
            "Order filled: {} {} {} @ ${}",
            side_str(side),
            quantity,
            instrument,
            price
        );

        self.orders.push(order.clone());
        Ok(order)
    }

    /// Get portfolio snapshot
    pub fn snapshot(&self) -> PortfolioSnapshot {
        let positions: Vec<Position> = self.positions.values().cloned().collect();
        let positions_value: Decimal = positions.iter().map(|p| p.market_value()).sum();
        let total_value = self.cash + positions_value;
        let total_pnl = total_value - self.initial_capital;
        let total_pnl_pct = if self.initial_capital.is_zero() {
            Decimal::ZERO
        } else {
            (total_pnl / self.initial_capital) * Decimal::ONE_HUNDRED
        };

        PortfolioSnapshot {
            timestamp: Utc::now(),
            cash: self.cash,
            positions_value,
            total_value,
            total_pnl,
            total_pnl_pct,
            positions,
            daily_return: None,
            sharpe_ratio: None,
            max_drawdown: None,
        }
    }
}

fn side_str(side: Side) -> &'static str {
    match side {
        Side::Buy => "BUY",
        Side::Sell => "SELL",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_paper_exchange_buy() {
        let mut exchange = PaperExchange::new(Decimal::new(10_000, 0));
        assert_eq!(exchange.cash(), Decimal::new(10_000, 0));

        let order = exchange
            .execute_market_order(
                "BTC",
                Side::Buy,
                Decimal::new(1, 1), // 0.1 BTC
                Decimal::new(70_000, 0),
                "Test buy",
            )
            .unwrap();

        assert_eq!(order.status, OrderStatus::Filled);
        assert_eq!(exchange.cash(), Decimal::new(3_000, 0)); // 10000 - 7000
        assert_eq!(exchange.positions().len(), 1);
    }

    #[test]
    fn test_paper_exchange_buy_sell_pnl() {
        let mut exchange = PaperExchange::new(Decimal::new(10_000, 0));

        // Buy 0.1 BTC at 70,000
        exchange
            .execute_market_order(
                "BTC",
                Side::Buy,
                Decimal::new(1, 1),
                Decimal::new(70_000, 0),
                "Buy",
            )
            .unwrap();

        // Price goes to 80,000
        exchange.update_price("BTC", Decimal::new(80_000, 0));

        // Sell at 80,000
        exchange
            .execute_market_order(
                "BTC",
                Side::Sell,
                Decimal::new(1, 1),
                Decimal::new(80_000, 0),
                "Sell",
            )
            .unwrap();

        // Should have 10,000 + 1,000 profit = 11,000
        assert_eq!(exchange.cash(), Decimal::new(11_000, 0));
        assert_eq!(exchange.positions().len(), 0);
    }

    #[test]
    fn test_insufficient_cash() {
        let mut exchange = PaperExchange::new(Decimal::new(100, 0));

        let result = exchange.execute_market_order(
            "BTC",
            Side::Buy,
            Decimal::new(1, 0), // 1 BTC
            Decimal::new(70_000, 0),
            "Too expensive",
        );

        assert!(result.is_err());
    }
}
